### PR TITLE
Add 'AggregateMetrics' rule to RuleEngine

### PR DESCRIPTION
**Pull request by sajjad-moradi**
## Description
The new rule checks the provided queries in the input of Rule Engine and suggests the value for 'AggregateMetrics' flag in table config. It looks at selection columns and if all of them are SUM function, the output flag is true, otherwise it's false.
## Testing Done
unit & integration tests

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRule.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.apache.pinot.controller.recommender.io.InputManager;
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.core.query.request.context.ExpressionContext;
+
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.HYBRID;
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.REALTIME;
+
+
+/**
+ * This rule checks the provided queries and suggest the value for 'AggregateMetrics' flag in table config.
```
Comment (jackjlli) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRule.java:33]: Nit: suggests

File: pinot-controller/src/test/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRuleTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    Map<String, Double> queryWithWeight = new HashMap<>();
+    for (String query : queries) {
+      queryWithWeight.put(query, 1.0);
+    }
+    input.setQueryWeightMap(queryWithWeight);
+    input.setTableType("Realtime");
+    input.init();
+    return input;
+  }
+}
```
Comment (jackjlli) [inline pinot-controller/src/test/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRuleTest.java:87]: Missing line at the end of this class.

**Review by jackjlli — APPROVED** [review]
Minor but LGTM.

File: pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRule.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+
+  private boolean shouldAggregate(InputManager inputManager) {
+    for (String query : inputManager.getParsedQueries()) {
+      for (ExpressionContext selectExpr : inputManager.getQueryContext(query).getSelectExpressions()) {
+        if (selectExpr.getType() != ExpressionContext.Type.FUNCTION
+            || !selectExpr.getFunction().getFunctionName().equalsIgnoreCase("SUM")) {
+          return false;
+        }
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRule.java:57]: Should we verify that column name part of SUM() is indeed a metric column name
Technically nothing prevents a user from doing aggregation function on a dimension column although it sounds less intuitive.

Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/AggregateMetricsRule.java:57]: That's a valid point. As you mentioned it's unlikely, but we better be safe. I'll add a check for column name being a metric.

**Comment by sajjad-moradi** [comment]
A few points about the updates on the PR:

1) I had to recursively go over the expression tree in each SUM function to find if the columns used in different branches of the expression tree are all metrics.

2) There was an invalid function name in the integration test (instead of SUM, I'd put SOME!!), but the test was passing because inputManager swallows invalid sql strings. I fixed the issues and made sure that new sql strings are valid and the rule gets executed properly.

**Review by siddharthteotia — APPROVED** [review]

**Pull request merged**